src/Avalonia.Vulkan/VulkanCommandBufferPool.cs

```diff
@@ -0,0 +1,175 @@
+        }
+
+        public VulkanCommandBuffer RentCommandBuffer()
+        {
+            return new(_device, this);
+        }
+
+        public void FreeUsedCommandBuffers()
+        {
+            lock (_usedCommandBuffers)
+            {
+                foreach (var usedCommandBuffer in _usedCommandBuffers) usedCommandBuffer.Dispose();
+
+                _usedCommandBuffers.Clear();
+            }
+        }
+
+        private void DisposeCommandBuffer(VulkanCommandBuffer commandBuffer)
+        {
+            lock (_usedCommandBuffers)
+            {
+                _usedCommandBuffers.Add(commandBuffer);
+            }
+        }
+
+        public class VulkanCommandBuffer : IDisposable
+        {
+            private readonly VulkanCommandBufferPool _commandBufferPool;
+            private readonly VulkanDevice _device;
+            private readonly Fence _fence;
+            public bool _isEnded;
+            public bool _isStarted;
+
+            internal unsafe VulkanCommandBuffer(VulkanDevice device, VulkanCommandBufferPool commandBufferPool)
+            {
+                _device = device;
+                _commandBufferPool = commandBufferPool;
+
+                ApiHandle = _commandBufferPool.AllocateCommandBuffer();
+
+                var fenceCreateInfo = new FenceCreateInfo()
+                {
+                    SType = StructureType.FenceCreateInfo,
+                    Flags = FenceCreateFlags.FenceCreateSignaledBit
+                };
+
+                device.Api.CreateFence(device.ApiHandle, fenceCreateInfo, null, out _fence);
+            }
+
+            public CommandBuffer ApiHandle { get; }
+
+            public unsafe void Dispose()
+            {
+                _device.Api.WaitForFences(_device.ApiHandle, 1, _fence, true, ulong.MaxValue);
+                _device.Api.FreeCommandBuffers(_device.ApiHandle, _commandBufferPool._commandPool, 1, ApiHandle);
+                _device.Api.DestroyFence(_device.ApiHandle, _fence, null);
+            }
+
+            public void BeginRecording()
+            {
+                if (!_isStarted)
+                {
+                    _isStarted = true;
+
+                    var beginInfo = new CommandBufferBeginInfo
+                    {
+                        SType = StructureType.CommandBufferBeginInfo,
+                        Flags = CommandBufferUsageFlags.CommandBufferUsageOneTimeSubmitBit
+                    };
+
+                    _device.Api.BeginCommandBuffer(ApiHandle, beginInfo);
+                }
+            }
+
+            public void EndRecording()
+            {
+                if (_isStarted && !_isEnded)
+                {
+                    _isEnded = true;
+
+                    _device.Api.EndCommandBuffer(ApiHandle);
+                }
+            }
+
+            public void Submit()
+            {
+                Submit(null, null, null, _fence);
+            }
+
+            public unsafe void Submit(
```

Comment:
I'm wondering if this API couldn't use `ReadOnlySpan<T>` instead of arrays. This would allow to skip allocations in most of the cases.

Reply:
Addressed